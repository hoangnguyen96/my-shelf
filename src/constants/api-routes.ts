export const API_ROUTES = {
  USER: "/user",
  BOOKS: "/books",
};
