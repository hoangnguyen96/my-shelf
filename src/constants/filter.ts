export enum TYPE_SEARCH {
  TITLE = "title",
  AUTHOR = "author",
}
