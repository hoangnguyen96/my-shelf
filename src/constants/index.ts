export * from "./input";
export * from "./navbar";
export * from "./routes";
export * from "./api-routes";
export * from "./validate";
export * from "./message";
export * from "./filter";
export * from "./time";
