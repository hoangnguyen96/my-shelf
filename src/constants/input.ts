export const PASSWORD = {
  HIDE: "Hide password",
  SHOW: "Show password",
};
