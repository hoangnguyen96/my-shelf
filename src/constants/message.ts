export const MESSAGES = {
  RESPONSE_ERROR: "The server is not responding. Please try again!",
  GET_ERROR: "The server does not respond. Retrieving data failed!",
  POST_ERROR: "The server does not respond. Please try again!",
  DELETE_ERROR: "Delete failed. Please try again!",
  EDIT_ERROR: "Edit failed. Please try again!",
  SUBMIT_ERROR: "Submit failed. Please try again!",
  UPDATE_ERROR: "Update failed. Please try again!",
  CREATE_ERROR: "Create data failed. Please try again!",
  ADD_USER_FAILED: "Failed to add user!",
  REGISTER_SUCCESS: "You have successfully created an account.",
  CREATE_FAILED: "Account creation failed!",

  // Response Api
  LOGIN_SUCCESS: "You have successfully logged in.",
  LOGIN_FAILED: "Login failed!",
  NETWORK_ERROR: "Network response was not ok!",
  EMAIL_EXISTS: "Email exists!",
  EMAIL_PASSWORD_INCORRECT: "Incorrect email or password!",
  EMAIL_PASSWORD_INVALID: "Email or password invalid.Please help to try again!",
  UPLOAD_FAILED: "Failed to upload image!",
  ERROR_UPLOAD: "Error uploading image!",

  // Errors
  LOGIN_NOTFOUND: "User not found or password missing",
  INVALID_CREDENTIALS: "Invalid credentials",

  // Validation
  FIELD_REQUIRED: "This field is required.",

  // Dynamic error messages
  FORMAT: (ariaLabel: string) => `${ariaLabel} does not follow format.`,

  // ConfirmPassword
  CONFIRM_PASSWORD: "Passwords do not match!",
};
