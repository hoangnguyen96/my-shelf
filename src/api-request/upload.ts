import { MESSAGES } from "@app/constants";

export const generateImageUpload = async (file: File) => {
  try {
    const formData = new FormData();
    formData.append("image", file);

    const response = await fetch(
      `https://api.imgbb.com/1/upload?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!response.ok) {
      throw new Error(MESSAGES.UPLOAD_FAILED);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw new Error(MESSAGES.ERROR_UPLOAD);
  }
};
