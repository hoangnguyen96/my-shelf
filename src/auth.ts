import NextAuth from "next-auth";
import { authConfig } from "./auth.config";
import Credentials from "next-auth/providers/credentials";
import { z } from "zod";
import bcrypt from "bcryptjs";
import { MESSAGES } from "./constants";
import { getUserByEmail } from "@app/features/dashboard/actions";
import { User } from "./models";

export const { handlers, auth, signIn, signOut } = NextAuth({
  ...authConfig,
  secret: process.env.AUTH_SECRET,
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: { label: "Email", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        if (!credentials) {
          return null;
        }

        const parsedCredentials = z
          .object({ email: z.string().email(), password: z.string().min(6) })
          .safeParse(credentials);

        if (parsedCredentials.success) {
          const { email, password } = parsedCredentials.data;
          const user = (await getUserByEmail(email)) as User;

          if (!user || !user.password) {
            throw new Error(MESSAGES.LOGIN_NOTFOUND);
          }

          const passwordsMatch = await bcrypt.compare(password, user.password);

          if (passwordsMatch)
            return {
              id: user.userId,
              name: user.username,
              isAdmin: user.isAdmin,
              image: user.avatar,
              email: user.email,
            };
        }

        throw new Error(MESSAGES.INVALID_CREDENTIALS);
      },
    }),
  ],
});
