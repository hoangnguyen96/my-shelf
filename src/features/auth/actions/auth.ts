"use server";

import { AuthError } from "next-auth";
import { MESSAGES } from "@app/constants";
import { User } from "@app/interface";
import { signIn, signOut } from "@app/auth";

export const authenticate = async (formData: Partial<User>) => {
  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      return MESSAGES.EMAIL_PASSWORD_INVALID;
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
