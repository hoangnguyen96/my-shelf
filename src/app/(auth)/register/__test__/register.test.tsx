import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { addUser, getUserByEmail } from "@app/api-request";
import { useRouter } from "next/navigation";
import RegisterPage from "../page";
import { MESSAGES } from "@app/constants";
import { authenticate } from "@app/actions/auth";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/api-request", () => ({
  addUser: jest.fn(),
  getUserByEmail: jest.fn(),
}));

jest.mock("@app/actions/auth", () => ({
  authenticate: jest.fn(),
}));

describe("Register", () => {
  const mockPush = jest.fn();
  const mockedAuthenticate = authenticate as jest.MockedFunction<
    typeof authenticate
  >;

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { asFragment } = render(<RegisterPage />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("Should handle add user failed", async () => {
    (addUser as jest.Mock).mockRejectedValue(
      new Error("Authentication failed after registration.")
    );
    await act(async () => {
      const { asFragment } = render(<RegisterPage />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("Should call onSubmit success with correct data when form is valid", async () => {
    (getUserByEmail as jest.Mock).mockResolvedValue(undefined);
    (addUser as jest.Mock).mockResolvedValue({ success: true });
    mockedAuthenticate.mockResolvedValue("credentials");
    const { getByTestId, getByPlaceholderText, getByLabelText } = render(
      <RegisterPage />
    );

    await act(async () => {
      fireEvent.change(getByPlaceholderText(/username.../i), {
        target: { value: "hung" },
      });
      fireEvent.change(getByPlaceholderText(/email.../i), {
        target: { value: "hung77@gmail.com" },
      });
      fireEvent.change(getByLabelText("Password"), {
        target: { value: "Abc789()()" },
      });
      fireEvent.change(getByLabelText("Confirm Password"), {
        target: { value: "Abc789()()" },
      });

      fireEvent.click(getByTestId("submit-register"));
    });

    await waitFor(() => {
      expect(addUser).toHaveBeenCalledWith({
        username: "hung",
        email: "hung77@gmail.com",
        password: expect.any(String),
        isAdmin: false,
        phone: "",
        bio: "",
        avatar: "https://i.ibb.co/SKHPQYq/avatar-default.webp",
        favorites: [],
        shelfBooks: [],
        userId: expect.any(String),
      });
      expect(mockPush).toHaveBeenCalledWith("/home");
    });
  });

  it("Should call onSubmit failed with correct data when form is valid", async () => {
    mockedAuthenticate.mockRejectedValue("credentials");
    (addUser as jest.Mock).mockRejectedValue(
      new Error(MESSAGES.ADD_USER_FAILED)
    );
    const { getByTestId, getByPlaceholderText, getByLabelText } = render(
      <RegisterPage />
    );

    await act(async () => {
      fireEvent.change(getByPlaceholderText(/username.../i), {
        target: { value: "hung" },
      });
      fireEvent.change(getByPlaceholderText(/email.../i), {
        target: { value: "hung77@gmail.com" },
      });
      fireEvent.change(getByLabelText("Password"), {
        target: { value: "Abc789()()" },
      });
      fireEvent.change(getByLabelText("Confirm Password"), {
        target: { value: "Abc789()()" },
      });

      fireEvent.click(getByTestId("submit-register"));
    });
  });

  it("Should show error when email already exists", async () => {
    (getUserByEmail as jest.Mock).mockResolvedValue([
      { email: "hung77@gmail.com" },
    ]);

    const { getByPlaceholderText, getByTestId, getByText } = render(
      <RegisterPage />
    );

    await act(async () => {
      fireEvent.change(getByPlaceholderText(/email.../i), {
        target: { value: "hung77@gmail.com" },
      });

      fireEvent.click(getByTestId("submit-register"));
    });
  });
});
