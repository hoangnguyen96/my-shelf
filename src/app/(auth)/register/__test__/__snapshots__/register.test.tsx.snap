// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Should handle add user failed 1`] = `<DocumentFragment />`;

exports[`Register Should render correctly snapshot 1`] = `<DocumentFragment />`;
